File: java/src/main/java/money_problem/domain/MissingExchangeRatesException.java
package money_problem.domain;

import java.util.List;
import java.util.stream.Collectors;

public class MissingExchangeRatesException extends Exception {
    public MissingExchangeRatesException(List<MissingExchangeRateException> missingExchangeRates) {
        super(missingExchangeRates.stream()
                .map(e -> String.format("[%s]", e.getMessage()))
                .collect(Collectors.joining(",", "Missing exchange rate(s): ", "")));
    }
}

File: scala/src/main/scala/domain/MissingExchangeRatesException.scala
package domain

sealed case class MissingExchangeRatesException(
    missingExchangeRates: Seq[MissingExchangeRateException]
) extends Exception(
      missingExchangeRates
        .map(e => s"[${e.getMessage}]")
        .mkString("Missing exchange rate(s): ", ",", "")
    ) {}
